postgres/datadog_checks/postgres/connections.py

```diff
@@ -92,8 +127,51 @@ def close_all_connections(self):
                     success = False
         return success
 
-    def _terminate_connection_unsafe(self, dbname):
-        db, _ = self._conns.pop(dbname, ConnectionWithTTL(None, None))
+    def done(self, dbname: str) -> None:
+        """
+        Mark a connection as done being used, so it can be evicted from the pool.
+        This function does not evict connections from the pool; it just marks them
+        as inactive.
+        done() can only be called on a connection in the same thread that the connection
+        was made.
+        """
+        with self._mu:
+            if self._conns[dbname].thread != threading.current_thread():
+                raise RuntimeError(
+                    "Cannot call done() for this dbname on this thread. Done() can only be called \
+                                   from the same thread the connection was made."
+                )
+
+            self._conns[dbname].active = False
+
+    def evict_lru(self) -> str:
+        """
+        Evict and close the inactive connection which was least recently used.
+        Return the dbname connection that was evicted.
+        """
+        with self._mu:
+            conns_list = dict(self._conns)
+            while True:
+                if not conns_list:
+                    break
+
+                eviction_candidate = self._get_lru(conns_list)
+                if self._conns[eviction_candidate].active:
+                    del conns_list[eviction_candidate]
+                    continue
+
+                # eviction candidate successfully found
+                self._terminate_connection_unsafe(eviction_candidate)
```

Comment:
Hmm, ok let's have a discussion about this. This will terminate a connection of something that could be an actively running query. How will evict_lru be used in the application code?

Reply:
Ok I reread the PR description. I think this logic is a tad risky. For instance, let's say an explain thread is trying to secure a connection. It could end up terminating a connection for an actively running query in the main integration. That would be bad because we'd be interrupting critical path functionality for a feature of secondary importance (explain plans)

Reply:
I think that is the problem I tried to solve by having this "active" flag that gets marked inactive by the `done()`.  
with the API change you suggested, the connection would get marked inactive once we leave the connection context. so, if a query is actively running it wouldn't pass the eviction check, and it wouldn't be terminated. i'm assuming the whole explain-plan-getting workflow for one DB can be run in between calling `get_connection` and `done()`. maybe this test illustrates that: https://github.com/DataDog/integrations-core/pull/14786/files#diff-7752da2e53447f1f5a1bf3cdb09cfa9875f4191fc80075f7d4f74b515bd77e93R313-R316
here i'm running "5 second long query"s on 30 threads, but when I try to evict it returns None because done() hasn't been called on those connections

Reply:
Ok thanks for clarifying! Check out https://github.com/DataDog/integrations-core/pull/14786#discussion_r1232817265 for how I read the code incorrectly.